transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java

```diff
@@ -60,10 +66,10 @@ protected AbstractNioUnsafe newUnsafe() {
     protected class NioByteUnsafe extends AbstractNioUnsafe {
 
         private void closeOnRead(ChannelPipeline pipeline) {
-            SelectionKey key = selectionKey();
-            setInputShutdown();
             if (isOpen()) {
                 if (Boolean.TRUE.equals(config().getOption(ChannelOption.ALLOW_HALF_CLOSURE))) {
+                    shutdownInput();
```

Comment:
@normanmaurer - previously `shutdownInput` set state internally and did not attempt to set the channel state or change the state of the underlying transport. Now that we do interact with the underlying transport we either close the underlying channel or shutdown channel (if the user enabled this behavior). NIO already [checks the input is open](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/sun/nio/ch/SocketChannelImpl.java#300) before reading anything. This is now consistent with EPOLL.